# UNSW-NB15 Network Intrusion Detection

## Task 1: Data Preparation

### Objective

Prepare the UNSW-NB15 network intrusion detection dataset for
exploratory data analysis and machine learning.

### Dataset

UNSW-NB15 contains network traffic records representing normal
network activity and multiple categories of network attacks.

### Task 1 Goals

- Understand the structure of the dataset
- Inspect features and data types
- Identify missing and invalid values
- Check for duplicate records
- Examine numerical and categorical variables
- Validate target variables
- Document data preparation assumptions
- Prepare the data for subsequent analysis and modeling

### Tools

- Python
- Pandas
- NumPy
- Matplotlib
- Seaborn
- Scikit-learn

In [41]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

print("Libraries imported successfully.")

Libraries imported successfully.


In [42]:
train_df = pd.read_csv("/UNSW_NB15_training-set.csv")

print("Training dataset loaded successfully.")

Training dataset loaded successfully.


In [43]:
train_df.shape

(175341, 45)

In [44]:
test_df = pd.read_csv("/UNSW_NB15_testing-set.csv")

print("Testing dataset loaded successfully.")

Testing dataset loaded successfully.


In [45]:
test_df.shape

(82332, 45)

In [46]:
features_df = pd.read_csv(
    "/UNSW-NB15_features.csv",
    encoding="cp1252"
)

print("Feature description loaded successfully.")

Feature description loaded successfully.


In [47]:
features_df.head()

,No.,Name,Type,Description
0,1,srcip,nominal,Source IP address
1,2,sport,integer,Source port number
2,3,dstip,nominal,Destination IP address
3,4,dsport,integer,Destination port number
4,5,proto,nominal,Transaction protocol


In [48]:
features_df.shape

(49, 4)

In [49]:
print("========== DATASET SHAPE ==========")
print("Training:", train_df.shape)
print("Testing :", test_df.shape)

print("\n========== DATA TYPES ==========")
print(train_df.dtypes)

print("\n========== MISSING VALUES ==========")
print(train_df.isnull().sum())

print("\n========== DUPLICATE ROWS ==========")
print(train_df.duplicated().sum())

print("\n========== INFINITE VALUES ==========")
numeric_data = train_df.select_dtypes(include=np.number)
print(np.isinf(numeric_data).sum().sum())

========== DATASET SHAPE ==========
Training: (175341, 45)
Testing : (82332, 45)

========== DATA TYPES ==========
id                     int64
dur                  float64
proto                 object
service               object
state                 object
spkts                  int64
dpkts                  int64
sbytes                 int64
dbytes                 int64
rate                 float64
sttl                   int64
dttl                   int64
sload                float64
dload                float64
sloss                  int64
dloss                  int64
sinpkt               float64
dinpkt               float64
sjit                 float64
djit                 float64
swin                   int64
stcpb                  int64
dtcpb                  int64
dwin                   int64
tcprtt               float64
synack               float64
ackdat               float64
smean                  int64
dmean                  int64
trans_depth            int64
response_body_l

In [50]:
print("========== LABEL DISTRIBUTION ==========")
print(train_df["label"].value_counts())

========== LABEL DISTRIBUTION ==========
label
1    119341
0     56000
Name: count, dtype: int64


In [51]:
print("========== ATTACK CATEGORY DISTRIBUTION ==========")
print(train_df["attack_cat"].value_counts(dropna=False))

========== ATTACK CATEGORY DISTRIBUTION ==========
attack_cat
Normal            56000
Generic           40000
Exploits          33393
Fuzzers           18184
DoS               12264
Reconnaissance    10491
Analysis           2000
Backdoor           1746
Shellcode          1133
Worms               130
Name: count, dtype: int64


In [52]:
features_df

,No.,Name,Type,Description
0,1,srcip,nominal,Source IP address
1,2,sport,integer,Source port number
2,3,dstip,nominal,Destination IP address
3,4,dsport,integer,Destination port number
4,5,proto,nominal,Transaction protocol
5,6,state,nominal,Indicates to the state and its dependent proto...
6,7,dur,Float,Record total duration
7,8,sbytes,Integer,Source to destination transaction bytes
8,9,dbytes,Integer,Destination to source transaction bytes
9,10,sttl,Integer,Source to destination time to live value


In [53]:
features_df.columns.tolist()

['No.', 'Name', 'Type ', 'Description']

In [54]:
for col in features_df.columns:
    print(repr(col))

'No.'
'Name'
'Type '
'Description'


In [55]:
features_df.columns = features_df.columns.str.strip()

In [56]:
features_df.columns.tolist()

['No.', 'Name', 'Type', 'Description']

In [57]:
features_df["Type"].value_counts()

,count
Type,
integer,20
Float,10
Integer,8
nominal,6
Timestamp,2
Binary,2
binary,1


In [58]:
feature_names = set(features_df["Name"].str.strip())
dataset_columns = set(train_df.columns)

print("Number of documented features:", len(feature_names))
print("Number of dataset columns:", len(dataset_columns))

print("\nDocumented features NOT in our dataset:")
print(sorted(feature_names - dataset_columns))

print("\nDataset columns NOT in the feature documentation:")
print(sorted(dataset_columns - feature_names))

Number of documented features: 49
Number of dataset columns: 45

Documented features NOT in our dataset:
['Dintpkt', 'Djit', 'Dload', 'Dpkts', 'Label', 'Ltime', 'Sintpkt', 'Sjit', 'Sload', 'Spkts', 'Stime', 'ct_src_ ltm', 'dmeansz', 'dsport', 'dstip', 'res_bdy_len', 'smeansz', 'sport', 'srcip']

Dataset columns NOT in the feature documentation:
['ct_src_ltm', 'dinpkt', 'djit', 'dload', 'dmean', 'dpkts', 'id', 'label', 'rate', 'response_body_len', 'sinpkt', 'sjit', 'sload', 'smean', 'spkts']


In [59]:
train_df.columns.tolist()

['id',
 'dur',
 'proto',
 'service',
 'state',
 'spkts',
 'dpkts',
 'sbytes',
 'dbytes',
 'rate',
 'sttl',
 'dttl',
 'sload',
 'dload',
 'sloss',
 'dloss',
 'sinpkt',
 'dinpkt',
 'sjit',
 'djit',
 'swin',
 'stcpb',
 'dtcpb',
 'dwin',
 'tcprtt',
 'synack',
 'ackdat',
 'smean',
 'dmean',
 'trans_depth',
 'response_body_len',
 'ct_srv_src',
 'ct_state_ttl',
 'ct_dst_ltm',
 'ct_src_dport_ltm',
 'ct_dst_sport_ltm',
 'ct_dst_src_ltm',
 'is_ftp_login',
 'ct_ftp_cmd',
 'ct_flw_http_mthd',
 'ct_src_ltm',
 'ct_srv_dst',
 'is_sm_ips_ports',
 'attack_cat',
 'label']

In [60]:
feature_info = features_df[["Name", "Type", "Description"]].copy()

feature_info["Name"] = feature_info["Name"].str.strip()

feature_info = feature_info.merge(
    train_df.dtypes.rename("Pandas_Dtype"),
    left_on="Name",
    right_index=True,
    how="left"
)

feature_info

,Name,Type,Description,Pandas_Dtype
0,srcip,nominal,Source IP address,NaN
1,sport,integer,Source port number,NaN
2,dstip,nominal,Destination IP address,NaN
3,dsport,integer,Destination port number,NaN
4,proto,nominal,Transaction protocol,object
5,state,nominal,Indicates to the state and its dependent proto...,object
6,dur,Float,Record total duration,float64
7,sbytes,Integer,Source to destination transaction bytes,int64
8,dbytes,Integer,Destination to source transaction bytes,int64
9,sttl,Integer,Source to destination time to live value,int64


In [61]:
feature_names = set(features_df["Name"].str.strip())
dataset_columns = set(train_df.columns)

print("Number of documented features:", len(feature_names))
print("Number of dataset columns:", len(dataset_columns))

print("\nDocumented features NOT in our dataset:")
print(sorted(feature_names - dataset_columns))

print("\nDataset columns NOT in the feature documentation:")
print(sorted(dataset_columns - feature_names))

Number of documented features: 49
Number of dataset columns: 45

Documented features NOT in our dataset:
['Dintpkt', 'Djit', 'Dload', 'Dpkts', 'Label', 'Ltime', 'Sintpkt', 'Sjit', 'Sload', 'Spkts', 'Stime', 'ct_src_ ltm', 'dmeansz', 'dsport', 'dstip', 'res_bdy_len', 'smeansz', 'sport', 'srcip']

Dataset columns NOT in the feature documentation:
['ct_src_ltm', 'dinpkt', 'djit', 'dload', 'dmean', 'dpkts', 'id', 'label', 'rate', 'response_body_len', 'sinpkt', 'sjit', 'sload', 'smean', 'spkts']


In [62]:
feature_info = features_df[["Name", "Type", "Description"]].copy()

feature_info["Name"] = feature_info["Name"].str.strip()

feature_info = feature_info.merge(
    train_df.dtypes.rename("Pandas_Dtype"),
    left_on="Name",
    right_index=True,
    how="left"
)

feature_info

,Name,Type,Description,Pandas_Dtype
0,srcip,nominal,Source IP address,NaN
1,sport,integer,Source port number,NaN
2,dstip,nominal,Destination IP address,NaN
3,dsport,integer,Destination port number,NaN
4,proto,nominal,Transaction protocol,object
5,state,nominal,Indicates to the state and its dependent proto...,object
6,dur,Float,Record total duration,float64
7,sbytes,Integer,Source to destination transaction bytes,int64
8,dbytes,Integer,Destination to source transaction bytes,int64
9,sttl,Integer,Source to destination time to live value,int64


In [63]:
train_df["state"].value_counts()

,count
state,
INT,82275
FIN,77825
CON,13152
REQ,1991
RST,83
ECO,12
PAR,1
URN,1
no,1


In [64]:
train_df["proto"].value_counts().head(20)

,count
proto,
tcp,79946
udp,63283
unas,12084
arp,2859
ospf,2595
sctp,1150
any,300
gre,225
ipv6,201


In [65]:
train_df["service"].value_counts()

,count
service,
-,94168
dns,47294
http,18724
smtp,5058
ftp-data,3995
ftp,3428
ssh,1302
pop3,1105
dhcp,94


In [66]:
pd.crosstab(
    train_df["state"],
    train_df["label"],
    normalize="index"
) * 100

label,0,1
state,,
CON,91.993613,8.006387
ECO,100.000000,0.000000
FIN,47.767427,52.232573
INT,6.946217,93.053783
PAR,100.000000,0.000000
REQ,46.459066,53.540934
RST,85.542169,14.457831
URN,100.000000,0.000000
no,100.000000,0.000000


In [67]:
pd.crosstab(
    train_df["proto"],
    train_df["label"],
    normalize="index"
) * 100

label,0,1
proto,,
3pc,0.0,100.0
a/n,0.0,100.0
aes-sp3-d,0.0,100.0
any,0.0,100.0
argus,0.0,100.0
...,...,...
wsn,0.0,100.0
xnet,0.0,100.0
xns-idp,0.0,100.0


In [68]:
pd.crosstab(
    train_df["service"],
    train_df["label"],
    normalize="index"
) * 100

label,0,1
service,,
-,38.773256,61.226744
dhcp,0.000000,100.000000
dns,15.843447,84.156553
ftp,35.530922,64.469078
ftp-data,63.879850,36.120150
http,28.562273,71.437727
irc,0.000000,100.000000
pop3,0.361991,99.638009
radius,16.666667,83.333333


In [69]:
proto_analysis = pd.crosstab(
    train_df["proto"],
    train_df["label"]
)

proto_analysis["total"] = proto_analysis.sum(axis=1)

proto_analysis["attack_percentage"] = (
    proto_analysis.get(1, 0) / proto_analysis["total"] * 100
)

proto_analysis.sort_values(
    "total",
    ascending=False
).head(20)

label,0,1,total,attack_percentage
proto,,,,
tcp,39121,40825,79946,51.065719
udp,13922,49361,63283,78.000411
unas,0,12084,12084,100.000000
arp,2859,0,2859,0.000000
ospf,64,2531,2595,97.533719
sctp,0,1150,1150,100.000000
any,0,300,300,100.000000
gre,0,225,225,100.000000
swipe,0,201,201,100.000000


In [70]:
service_analysis = pd.crosstab(
    train_df["service"],
    train_df["label"]
)

service_analysis["total"] = service_analysis.sum(axis=1)

service_analysis["attack_percentage"] = (
    service_analysis.get(1, 0) / service_analysis["total"] * 100
)

service_analysis.sort_values(
    "total",
    ascending=False
)

label,0,1,total,attack_percentage
service,,,,
-,36512,57656,94168,61.226744
dns,7493,39801,47294,84.156553
http,5348,13376,18724,71.437727
smtp,1579,3479,5058,68.782127
ftp-data,2552,1443,3995,36.120150
ftp,1218,2210,3428,64.469078
ssh,1291,11,1302,0.844854
pop3,4,1101,1105,99.638009
dhcp,0,94,94,100.000000


In [71]:
train_df["label"].value_counts()

,count
label,
1,119341
0,56000


In [73]:
train_df["attack_cat"].value_counts(dropna=False)

,count
attack_cat,
Normal,56000
Generic,40000
Exploits,33393
Fuzzers,18184
DoS,12264
Reconnaissance,10491
Analysis,2000
Backdoor,1746
Shellcode,1133


In [74]:
pd.crosstab(train_df["attack_cat"], train_df["label"])

label,0,1
attack_cat,,
Analysis,0,2000
Backdoor,0,1746
DoS,0,12264
Exploits,0,33393
Fuzzers,0,18184
Generic,0,40000
Normal,56000,0
Reconnaissance,0,10491
Shellcode,0,1133


In [75]:
numeric_summary = train_df.describe().T

numeric_summary

,count,mean,std,min,25%,50%,75%,max
id,175341.0,8.767100e+04,5.061673e+04,1.0,43836.000000,87671.000000,1.315060e+05,1.753410e+05
dur,175341.0,1.359389e+00,6.480249e+00,0.0,0.000008,0.001582,6.680690e-01,5.999999e+01
spkts,175341.0,2.029866e+01,1.368876e+02,1.0,2.000000,2.000000,1.200000e+01,9.616000e+03
dpkts,175341.0,1.896959e+01,1.102583e+02,0.0,0.000000,2.000000,1.000000e+01,1.097400e+04
sbytes,175341.0,8.844844e+03,1.747656e+05,28.0,114.000000,430.000000,1.418000e+03,1.296523e+07
dbytes,175341.0,1.492892e+04,1.436542e+05,0.0,0.000000,164.000000,1.102000e+03,1.465555e+07
rate,175341.0,9.540619e+04,1.654010e+05,0.0,32.786140,3225.806520,1.250000e+05,1.000000e+06
sttl,175341.0,1.795470e+02,1.029400e+02,0.0,62.000000,254.000000,2.540000e+02,2.550000e+02
dttl,175341.0,7.960957e+01,1.105069e+02,0.0,0.000000,29.000000,2.520000e+02,2.540000e+02
sload,175341.0,7.345403e+07,1.883574e+08,0.0,13053.338870,879674.750000,8.888889e+07,5.988000e+09


In [76]:
binary_columns = [
    "is_ftp_login",
    "is_sm_ips_ports",
    "label"
]

for col in binary_columns:
    print(f"\n{col}")
    print(train_df[col].value_counts().sort_index())


is_ftp_login
is_ftp_login
0    172774
1      2545
2         6
4        16
Name: count, dtype: int64

is_sm_ips_ports
is_sm_ips_ports
0    172579
1      2762
Name: count, dtype: int64

label
label
0     56000
1    119341
Name: count, dtype: int64


In [77]:
suspicious_columns = [
    "spkts",
    "dpkts",
    "sbytes",
    "dbytes",
    "sloss",
    "dloss",
    "trans_depth",
    "response_body_len"
]

for col in suspicious_columns:
    print(f"\n===== {col} =====")
    print(train_df[col].describe())


===== spkts =====
count    175341.000000
mean         20.298664
std         136.887597
min           1.000000
25%           2.000000
50%           2.000000
75%          12.000000
max        9616.000000
Name: spkts, dtype: float64

===== dpkts =====
count    175341.000000
mean         18.969591
std         110.258271
min           0.000000
25%           0.000000
50%           2.000000
75%          10.000000
max       10974.000000
Name: dpkts, dtype: float64

===== sbytes =====
count    1.753410e+05
mean     8.844844e+03
std      1.747656e+05
min      2.800000e+01
25%      1.140000e+02
50%      4.300000e+02
75%      1.418000e+03
max      1.296523e+07
Name: sbytes, dtype: float64

===== dbytes =====
count    1.753410e+05
mean     1.492892e+04
std      1.436542e+05
min      0.000000e+00
25%      0.000000e+00
50%      1.640000e+02
75%      1.102000e+03
max      1.465555e+07
Name: dbytes, dtype: float64

===== sloss =====
count    175341.000000
mean          4.953000
std          66.005059


In [78]:
negative_counts = (train_df.select_dtypes(include=np.number) < 0).sum()

negative_counts[negative_counts > 0]

,0


In [79]:
nunique = train_df.nunique()

nunique.sort_values()

,0
is_sm_ips_ports,2
label,2
is_ftp_login,4
ct_ftp_cmd,4
ct_state_ttl,5
dttl,6
dwin,7
state,9
attack_cat,10
ct_flw_http_mthd,11


In [80]:
train_df[train_df["is_ftp_login"].isin([2, 4])][
    ["is_ftp_login", "service", "proto", "state", "attack_cat", "label"]
]

,is_ftp_login,service,proto,state,attack_cat,label
13114,2,ftp,tcp,FIN,Normal,0
13130,2,ftp,tcp,FIN,Normal,0
48027,2,ftp,tcp,FIN,Fuzzers,1
48224,2,ftp,tcp,FIN,Fuzzers,1
48306,2,ftp,tcp,FIN,Fuzzers,1
48439,2,ftp,tcp,FIN,Fuzzers,1
81497,4,ftp,tcp,CON,Exploits,1
81522,4,ftp,tcp,FIN,Exploits,1
82973,4,ftp,tcp,CON,Exploits,1
86897,4,ftp,tcp,CON,Exploits,1


In [81]:
train_df["ct_ftp_cmd"].value_counts().sort_index()

,count
ct_ftp_cmd,
0,172774
1,2545
2,6
4,16


In [82]:
train_df[train_df["ct_ftp_cmd"] > 1][
    ["ct_ftp_cmd", "is_ftp_login", "service", "attack_cat", "label"]
]

,ct_ftp_cmd,is_ftp_login,service,attack_cat,label
13114,2,2,ftp,Normal,0
13130,2,2,ftp,Normal,0
48027,2,2,ftp,Fuzzers,1
48224,2,2,ftp,Fuzzers,1
48306,2,2,ftp,Fuzzers,1
48439,2,2,ftp,Fuzzers,1
81497,4,4,ftp,Exploits,1
81522,4,4,ftp,Exploits,1
82973,4,4,ftp,Exploits,1
86897,4,4,ftp,Exploits,1


In [83]:
negative_counts = (
    train_df.select_dtypes(include=np.number) < 0
).sum()

negative_counts[negative_counts > 0]

,0


In [84]:
pd.crosstab(
    train_df["is_ftp_login"],
    train_df["ct_ftp_cmd"]
)

ct_ftp_cmd,0,1,2,4
is_ftp_login,,,,
0,172774,0,0,0
1,0,2545,0,0
2,0,0,6,0
4,0,0,0,16


In [85]:
train_df[
    (train_df["is_ftp_login"] > 1) |
    (train_df["ct_ftp_cmd"] > 1)
][
    ["is_ftp_login", "ct_ftp_cmd", "service", "attack_cat", "label"]
].value_counts()

is_ftp_login  ct_ftp_cmd  service  attack_cat  label
4             4           ftp      Exploits    1        16
2             2           ftp      Fuzzers     1         4
                                   Normal      0         2
Name: count, dtype: int64

In [86]:
print("Same columns:", list(train_df.columns) == list(test_df.columns))

Same columns: True


In [87]:
dtype_comparison = pd.DataFrame({
    "train_dtype": train_df.dtypes,
    "test_dtype": test_df.dtypes
})

dtype_comparison[
    dtype_comparison["train_dtype"] != dtype_comparison["test_dtype"]
]

,train_dtype,test_dtype


In [88]:
categorical_columns = ["proto", "service", "state", "attack_cat"]

for col in categorical_columns:
    train_categories = set(train_df[col].dropna().unique())
    test_categories = set(test_df[col].dropna().unique())

    print(f"\n===== {col} =====")
    print("Only in train:", sorted(train_categories - test_categories))
    print("Only in test :", sorted(test_categories - train_categories))


===== proto =====
Only in train: ['icmp', 'rtp']
Only in test : []

===== service =====
Only in train: []
Only in test : []

===== state =====
Only in train: ['ECO', 'PAR', 'URN', 'no']
Only in test : ['ACC', 'CLO']

===== attack_cat =====
Only in train: []
Only in test : []


In [89]:
for col in train_df.columns:
    if col != col.strip():
        print("Column with whitespace:", repr(col))

In [90]:
print("Training columns:")
print(train_df.columns.tolist())

print("\nTesting columns:")
print(test_df.columns.tolist())

Training columns:
['id', 'dur', 'proto', 'service', 'state', 'spkts', 'dpkts', 'sbytes', 'dbytes', 'rate', 'sttl', 'dttl', 'sload', 'dload', 'sloss', 'dloss', 'sinpkt', 'dinpkt', 'sjit', 'djit', 'swin', 'stcpb', 'dtcpb', 'dwin', 'tcprtt', 'synack', 'ackdat', 'smean', 'dmean', 'trans_depth', 'response_body_len', 'ct_srv_src', 'ct_state_ttl', 'ct_dst_ltm', 'ct_src_dport_ltm', 'ct_dst_sport_ltm', 'ct_dst_src_ltm', 'is_ftp_login', 'ct_ftp_cmd', 'ct_flw_http_mthd', 'ct_src_ltm', 'ct_srv_dst', 'is_sm_ips_ports', 'attack_cat', 'label']

Testing columns:
['id', 'dur', 'proto', 'service', 'state', 'spkts', 'dpkts', 'sbytes', 'dbytes', 'rate', 'sttl', 'dttl', 'sload', 'dload', 'sloss', 'dloss', 'sinpkt', 'dinpkt', 'sjit', 'djit', 'swin', 'stcpb', 'dtcpb', 'dwin', 'tcprtt', 'synack', 'ackdat', 'smean', 'dmean', 'trans_depth', 'response_body_len', 'ct_srv_src', 'ct_state_ttl', 'ct_dst_ltm', 'ct_src_dport_ltm', 'ct_dst_sport_ltm', 'ct_dst_src_ltm', 'is_ftp_login', 'ct_ftp_cmd', 'ct_flw_http_mthd', 

In [91]:
train_clean = train_df.copy()
test_clean = test_df.copy()

In [92]:
print("Original training shape :", train_df.shape)
print("Cleaned training shape  :", train_clean.shape)

print("Original testing shape  :", test_df.shape)
print("Cleaned testing shape   :", test_clean.shape)

Original training shape : (175341, 45)
Cleaned training shape  : (175341, 45)
Original testing shape  : (82332, 45)
Cleaned testing shape   : (82332, 45)


In [93]:
train_clean.columns = train_clean.columns.str.strip()
test_clean.columns = test_clean.columns.str.strip()

In [94]:
print("Training columns:", len(train_clean.columns))
print("Testing columns :", len(test_clean.columns))

print("Column names identical:",
      list(train_clean.columns) == list(test_clean.columns))

Training columns: 45
Testing columns : 45
Column names identical: True


In [95]:
for col in train_clean.columns:
    if col != col.strip():
        print("Whitespace found:", repr(col))

In [96]:
missing_values = train_clean.isnull().sum()

missing_values[missing_values > 0]

,0


In [97]:
missing_values_test = test_clean.isnull().sum()

missing_values_test[missing_values_test > 0]

,0


In [98]:
print("Duplicate rows in training:", train_clean.duplicated().sum())
print("Duplicate rows in testing :", test_clean.duplicated().sum())

Duplicate rows in training: 0
Duplicate rows in testing : 0


In [99]:
numeric_train = train_clean.select_dtypes(include=np.number)

print(
    "Infinite values in training:",
    np.isinf(numeric_train).sum().sum()
)

Infinite values in training: 0


In [100]:
numeric_test = test_clean.select_dtypes(include=np.number)

print(
    "Infinite values in testing:",
    np.isinf(numeric_test).sum().sum()
)

Infinite values in testing: 0


In [101]:
categorical_columns = train_clean.select_dtypes(
    include="object"
).columns

for col in categorical_columns:
    print(f"\n===== {col} =====")
    print("Values with leading/trailing whitespace:",
          train_clean[col].astype(str).str.strip().ne(
              train_clean[col].astype(str)
          ).sum())


===== proto =====
Values with leading/trailing whitespace: 0

===== service =====
Values with leading/trailing whitespace: 0

===== state =====
Values with leading/trailing whitespace: 0

===== attack_cat =====
Values with leading/trailing whitespace: 0


In [102]:
feature_mapping = pd.DataFrame({
    "Dataset_Column": train_clean.columns
})

feature_mapping

,Dataset_Column
0,id
1,dur
2,proto
3,service
4,state
5,spkts
6,dpkts
7,sbytes
8,dbytes
9,rate


In [103]:
def normalize_name(name):
    return (
        str(name)
        .strip()
        .lower()
        .replace(" ", "")
        .replace("_", "")
    )

documentation_names = {
    normalize_name(name): name
    for name in features_df["Name"]
}

dataset_names = {
    normalize_name(name): name
    for name in train_clean.columns
}

documentation_not_in_dataset = [
    documentation_names[name]
    for name in documentation_names
    if name not in dataset_names
]

dataset_not_in_documentation = [
    dataset_names[name]
    for name in dataset_names
    if name not in documentation_names
]

print("Documentation fields without normalized dataset match:")
print(documentation_not_in_dataset)

print("\nDataset columns without normalized documentation match:")
print(dataset_not_in_documentation)

Documentation fields without normalized dataset match:
['srcip', 'sport', 'dstip', 'dsport', 'smeansz', 'dmeansz', 'res_bdy_len', 'Stime', 'Ltime', 'Sintpkt', 'Dintpkt']

Dataset columns without normalized documentation match:
['id', 'rate', 'sinpkt', 'dinpkt', 'smean', 'dmean', 'response_body_len']


In [104]:
feature_info = features_df[["Name", "Type", "Description"]].copy()

feature_info["Name"] = feature_info["Name"].str.strip()

feature_info["Dataset_Column"] = feature_info["Name"].map(
    lambda x: dataset_names.get(normalize_name(x))
)
feature_info

,Name,Type,Description,Dataset_Column
0,srcip,nominal,Source IP address,None
1,sport,integer,Source port number,None
2,dstip,nominal,Destination IP address,None
3,dsport,integer,Destination port number,None
4,proto,nominal,Transaction protocol,proto
5,state,nominal,Indicates to the state and its dependent proto...,state
6,dur,Float,Record total duration,dur
7,sbytes,Integer,Source to destination transaction bytes,sbytes
8,dbytes,Integer,Destination to source transaction bytes,dbytes
9,sttl,Integer,Source to destination time to live value,sttl


In [105]:
print("Unique IDs:", train_clean["id"].nunique())
print("Rows:", len(train_clean))

Unique IDs: 175341
Rows: 175341


## Identifier Analysis

The `id` column contains 175,341 unique values for 175,341
training records, indicating that it functions as a unique record
identifier.

The column was retained in the prepared dataset to preserve the
original record structure. It will not be treated as a predictive
feature during subsequent machine-learning modeling.

In [106]:
# Final data quality validation

print("========== FINAL VALIDATION ==========")

print("\nTraining shape:", train_clean.shape)
print("Testing shape :", test_clean.shape)

print("\nMissing values:")
print("Training:", train_clean.isnull().sum().sum())
print("Testing :", test_clean.isnull().sum().sum())

print("\nDuplicate rows:")
print("Training:", train_clean.duplicated().sum())
print("Testing :", test_clean.duplicated().sum())

print("\nInfinite values:")
print(
    "Training:",
    np.isinf(train_clean.select_dtypes(include=np.number)).sum().sum()
)
print(
    "Testing :",
    np.isinf(test_clean.select_dtypes(include=np.number)).sum().sum()
)

print("\nColumns identical:",
      list(train_clean.columns) == list(test_clean.columns))

print("\nData types identical:",
      train_clean.dtypes.equals(test_clean.dtypes))

========== FINAL VALIDATION ==========

Training shape: (175341, 45)
Testing shape : (82332, 45)

Missing values:
Training: 0
Testing : 0

Duplicate rows:
Training: 0
Testing : 0

Infinite values:
Training: 0
Testing : 0

Columns identical: True

Data types identical: True


## Final Data Preparation Summary

The UNSW-NB15 training and testing datasets were systematically
audited before preparation.

### Data quality findings

- No missing values were found.
- No duplicate rows were found.
- No infinite numerical values were found.
- No negative numerical values were found.
- Training and testing datasets contain the same 45 columns.
- Training and testing columns have matching data types.
- `id` contains a unique value for every record and was treated as
  a record identifier rather than a predictive feature.
- Categorical variables were inspected for their distributions and
  train/test category differences.
- Some categories occur only in one dataset split, particularly
  within `proto` and `state`. These categories were retained rather
  than removed.
- The `is_ftp_login` column is documented as binary, but the prepared
  dataset contains values 0, 1, 2, and 4. The values 2 and 4 occur
  only in FTP records and correspond exactly to `ct_ftp_cmd`. These
  values were retained because the supplied documentation does not
  establish a justified recoding rule.

### Preparation decisions

No unnecessary rows or observations were removed. Since the dataset
contained no missing, duplicate, infinite, or negative values, no
imputation or deletion was required.

Column names were standardized by removing leading and trailing
whitespace.

The supplied feature documentation contains 49 documented fields,
while the prepared train/test datasets contain 45 columns. Naming
differences were documented rather than arbitrarily changing the
dataset schema. Fields not available in the prepared datasets were
not reconstructed or imputed.

The resulting datasets preserve the available source information
while documenting the identified schema and data-definition
differences.

## Assumptions and Limitations

1. The prepared UNSW-NB15 train/test CSV files were treated as the
   source data for this analysis.

2. The supplied feature-description file was used as supporting
   documentation for interpreting the dataset columns.

3. Differences between documentation names and dataset column names
   were treated as schema/naming differences unless there was
   sufficient evidence to identify them as data-quality errors.

4. Rare categorical values were retained because rarity alone does
   not establish that a value is invalid.

5. Extreme numerical values were retained because their magnitude
   alone does not establish that they are erroneous network
   measurements.

6. The `id` column was retained for record identification but should
   not be used as a predictive feature in subsequent machine-learning
   models.

7. The `is_ftp_login` values 2 and 4 were retained because the
   available documentation does not provide enough information to
   justify converting them to another value.